informing/internal/services/informing_service.go

```diff
@@ -1,50 +1,67 @@
 package services
 
 import (
+	"informing-service/internal/models"
 	"log"
 )
 
 type (
 	IEmailSender interface {
 		SendInforming(subscriptions []string, rate float64)
 	}
-	SubscriptionClientInterface interface {
-		ListSubscribed() ([]string, error)
+
+	SubscriptionRepositoryInterface interface {
+		Create(email string) (*models.Subscription, error)
+		ListSubscribed() ([]models.Subscription, error)
+		Update(subscription models.Subscription) (*models.Subscription, error)
 	}
-	RateClientInterface interface {
-		FetchRate() (*float64, error)
+
+	RateRepositoryInterface interface {
+		Create(rate models.Rate) (*models.Rate, error)
+		GetLatest() (*models.Rate, error)
 	}
+
 	InformingServiceInterface interface {
 		SendEmails()
 	}
+
 	InformingService struct {
-		EmailSender        IEmailSender
-		subscriptionClient SubscriptionClientInterface
-		rateClient         RateClientInterface
+		EmailSender            IEmailSender
+		subscriptionRepository SubscriptionRepositoryInterface
+		rateRepository         RateRepositoryInterface
 	}
 )
 
-func NewInformingService(subscriptionClient SubscriptionClientInterface, rateClient RateClientInterface, sender IEmailSender) *InformingService {
+func NewInformingService(subscriptionRepository SubscriptionRepositoryInterface,
+	rateRepository RateRepositoryInterface,
+	sender IEmailSender) *InformingService {
 	return &InformingService{
-		EmailSender:        sender,
-		subscriptionClient: subscriptionClient,
-		rateClient:         rateClient,
+		EmailSender:            sender,
+		rateRepository:         rateRepository,
+		subscriptionRepository: subscriptionRepository,
 	}
 }
 
 func (s *InformingService) SendEmails() {
```

Comment:
We still need to think (maybe in the next MR) about optimising the load, introducing RPS limits for sending emails. Right now, you're sending emails to all subscribers in parallel at the same time.

Reply:
I have reworked this thing in the next homework. The new version pushes command to the queue to send email and consumer processes each message one by one